Create/Update tenant with ReCAPTCHA Config

Xiaoshouzi-gh opened the pull request:
Support reCaptcha config /create update on tenants.
     - Updated reCaptcha config type align with backend API.
     - Support create and update tenants with reCaptcha config.
     - Added reCaptcha unit tests on tenants.

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here. 
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help 
    us make Firebase APIs better, please propose your change in an issue so that we 
    can discuss it together.

renkelvin commented on src/auth/tenant.ts:
```diff
@@ -21,7 +21,7 @@ import { AuthClientErrorCode, FirebaseAuthError } from '../utils/error';
   EmailSignInConfig, EmailSignInConfigServerRequest, MultiFactorAuthServerConfig,
   MultiFactorConfig, validateTestPhoneNumbers, EmailSignInProviderConfig,
-  MultiFactorAuthConfig,
+  MultiFactorAuthConfig, RecaptchaConfigAuth, RecaptchaConfig
```
Change to *AuthConfig to match MultiFactorAuthConfig?

Xiaoshouzi-gh replied:
Done.

renkelvin commented on test/unit/auth/tenant.spec.ts:
```diff
@@ -79,6 +79,55 @@ describe('Tenant', () => {
     },
   };
 
+  const serverRequestWithRecaptcha: TenantServerResponse = {
```
serverResponse?

Xiaoshouzi-gh replied:
Done

renkelvin reviewed (commented):
Looks good to me. Please add Lahiru for approval.

lahirumaramba reviewed (changes requested):
Thank you, Liubin!
I added some comments and questions. Didn't get a chance to dive into unit tests. I will take another look later. Thanks!

lahirumaramba commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
+  key: string;
 }
 
 export interface RecaptchaConfig {
```
Please add doc strings here. Something like:
```
/**
 * The request interface for updating a ReCAPTCHA Config.
 */
```

lahirumaramba commented on etc/firebase-admin.auth.api.md:
```diff
@@ -264,6 +264,34 @@ export interface ProviderIdentifier {
+export type RecaptchaAction = 'BLOCK';
+
+// @public (undocumented)
+export interface RecaptchaConfig {
```
Please document this where you `export interface RecaptchaConfig`. API Extractor will pick it up from there. Thanks!

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
   * The enforcement state of email password provider.
   */
- emailPasswordEnforcementState?: RecaptchaProviderEnforcementState;
+  emailPasswordEnforcementState?: RecaptchaProviderEnforcementState;
```
the identation of the code is correct here, but missing a space for the documentation.  Updated.

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
+          '"RecaptchaConfig.emailPasswordEnforcementState" must be either "OFF", "AUDIT" or "ENFORCE".',
+        );
+      }
+    }
```
The reason for this separation is to align with the backend error-code and keep consistency with the existing error message.  E.g. empty string is an INVALID_ARGUMENT but invalid string is an INVALID_CONFIG.

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
+  public readonly managedRules?: RecaptchaManagedRule[];
+  public readonly recaptchaKeys?: RecaptchaKey[];
+
+  constructor(recaptchaConfig: RecaptchaConfig | undefined) {
```
Remove the double undefined check from auth-config part.  Tenant request is still valid when recaptchaConfig is undefined.

Xiaoshouzi-gh commented on src/auth/tenant.ts:
```diff
@@ -152,6 +160,9 @@ export class Tenant {
       request.testPhoneNumbers = tenantOptions.testPhoneNumbers ?? {};
     }
+    if (typeof tenantOptions.recaptchaConfig !== 'undefined') {
+      request.recaptchaConfig = tenantOptions.recaptchaConfig;
```
Please also refer to the reply for the next comment.  Besides the logic above logic, there is also another benefit not calling `buildServerRequest` is the validation part for both `MFA` and `EmailSignInConfig` got called unnecessarily twice in this constructor.

lahirumaramba replied:
Good point! Thanks!

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
+      // Validate each rule of the array
+      options.managedRules.forEach((managedRule) => {
+        ManagedRuleAuth.validate(managedRule);
+      });
```
Updated.

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1480,6 +1480,54 @@ export interface RecaptchaManagedRule {
  action?: RecaptchaAction;
 }
 
+export class ManagedRuleAuth implements RecaptchaManagedRule{
```
This class has been removed. Not applicable anymore

Xiaoshouzi-gh commented on src/auth/auth-config.ts:
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
+    }
+
+    if (typeof json.emailPasswordEnforcementState === 'undefined') {
+      delete json.emailPasswordEnforcementState;
```
`"emailPasswordEnforcementState": [undefined]` will be shown in the response.  This is also following the existing `multiFactorConfig` and `emailSignInConfig` class `toJSON()` method behavior.

Xiaoshouzi-gh commented:
> Thank you, Liubin! I added some comments and questions. Didn't get a chance to dive into unit tests. I will take another look later. Thanks!

Thank you @lahirumaramba for the review.  Addressed majority comments.  Resolved the trivial changes and left comments on the ones needed.

lahirumaramba commented on src/auth/tenant.ts:
```diff
@@ -185,6 +199,7 @@ export class Tenant {
       anonymousSignInEnabled: true,
       multiFactorConfig: true,
       testPhoneNumbers: true,
+      recaptchaConfig: true
```
```suggestion
      recaptchaConfig: true,
```

lahirumaramba reviewed (approved):
Thank you! Just a few nits but otherwise LGTM!
